compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -749,6 +758,130 @@ object desugar {
     Bind(name, Ident(nme.WILDCARD)).withPos(tree.pos)
   }
 
+  def decomposeTypePattern(tree: Tree)(implicit ctx: Context): (Tree, List[TypeDef]) = {
+    val bindingsBuf = new ListBuffer[TypeDef]
+    val elimTypeDefs = new untpd.UntypedTreeMap {
+      override def transform(tree: Tree)(implicit ctx: Context) = tree match {
+        case tree: TypeDef =>
+          bindingsBuf += tree
+          Ident(tree.name).withPos(tree.pos)
+        case _ =>
+          super.transform(tree)
+      }
+    }
+    (elimTypeDefs.transform(tree), bindingsBuf.toList)
+  }
+
+  private val collectNames = new untpd.UntypedTreeAccumulator[ListBuffer[String]] {
+    override def apply(buf: ListBuffer[String], tree: Tree)(implicit ctx: Context): ListBuffer[String] = tree match {
+      case Ident(name) =>
+        buf += name.toString
+      case Select(qual, name) =>
+        apply(buf, qual) += name.toString
+      case TypeDef(name, rhs) =>
+        apply(buf += "type" += name.toString, rhs)
+      case Apply(fn, _) =>
+        apply(buf, fn)
+      case _ =>
+        foldOver(buf, tree)
+    }
+  }
+
+  private def extensionName(ext: Extension)(implicit ctx: Context): TypeName = {
+    var buf = collectNames(new ListBuffer[String] += "extend", ext.extended)
+    if (ext.impl.parents.nonEmpty)
+      buf = collectNames(buf += "implements", ext.impl.parents)
+    val ids = buf.toList.map(_.filter(isIdentifierPart)).filter(!_.isEmpty)
+    ids.mkString("_").toTypeName
+  }
+
+  /**     extend <type-pattern> <params> implements <parents> { <body>} }
+   *   ->
+   *      implicit class <extension-name> <type-params> ($this: <extended>) <combined-params>
+   *      extends <parents> { <body1> }
+   *
+   *  where
+   *
+   *    <extension-name> = concatenation of all alphanumeric characters between and including `extend` and `{`,
```

Comment:
Let's consider the impact on binary compatibility. Are there valid changes to the signature, which result in a new name,  but which should not affect binary compatibility? Generally, I think it's important to give the programmer the ability to specify a name explicitly.

Reply:
One easy example: exchange a fully-qualified for a relative name or vice versa.
```scala
augment List[String] {
```
would be binary-incompatible with
```scala
augment List[java.lang.String] {
```
under Martin's naming proposal.

Also, renaming a `type T` parameter into `type U`.

Plenty more like that.

Reply:
Could use fully qualified names & a positional naming scheme for type params?

Reply:
Sorry, I think what I meant to say was type patterns? The `type T` and `type U` that @sjrd mentioned above. I think you already have a double definition if you have an extension for `type T` and `type U` that share the same method name.

Reply:
>  I think you already have a double definition if you have an extension for type T and type U that share the same method name.

No that's precisely it -  you want to enable multiple extends of the same type. So you should not get a double definition if you change the parameter type.